### PR TITLE
Or, you know, just do all the approaches for #266

### LINKED ISSUE #266: Add a `?skip_unchanged_versions=true|false` querysting argument to /imports
## Background

When importing data from Internet Archive (or any source, really), we only know about data in the timeframe we are examining. If a data source (like Internet Archive) doesn’t differentiate between versions that are the same and versions that are different from the ones that precede them, we could wind up importing a version to our database that is no different from the latest version we already have. This is undesirable when we are *generally* trying to only store versions that differ from the version before them (in the long run, that might be worth revisiting, but for now, it’s what we’re doing). We don’t make any guarantees about that, but we try to make sure it’s true in most cases.

For example, suppose our database has versions of the page `http://ugabooga.com/whiz-what-wow.html` like:

| capture_time | hash |
| ------------ | ---- |
| 2018-01-03   | abc  |
| 2018-01-05   | def  |
| 2018-01-20   | ghi  |

And Internet Archive captures a snapshot of that page every day. If the page never changes after 2018-01-20, but we run a script on 2018-01-21 to import new versions, we’ll bet back a version record from Internet Archive like:

| capture_time | hash |
| ------------ | ---- |
| 2018-01-21   | ghi  |

It’s exactly the same content as the latest version we already have (it has the same hash), but looks like a new record because it was captured on a newer date.

At current, we will store this version, which means our API will show two consecutive versions that are exactly the same :(

## Proposal

I propose we add a querystring argument to `POST /imports` called `?skip_unchanged_versions` with a boolean value. If set to `true`, the importer would not import versions of a page that have the same hash as the version captured immediately before them (*from the same source*).

The default value for it would be `false` — matching existing behavior, but also playing it safe. If `true`, we have to guarantee that the versions in the import request *are in capture_time order* (otherwise we might skip them incorrectly).

## Alternatives

We might start storing all snapshots of a page (regardless of whether they are different from the previous), but add a special field named `changed` (or `is_changed` or `is_different` or something similar). That field would be a boolean indicating whether that version’s hash matches the previous version’s hash, and we’d use it when querying to determine whether to show a version.

This is a lot more complicated than simply not importing things:

- We have to make sure the special database field stays up-to-date and accurate whenever adjacent versions are added
- We have to make sure all our queries correctly check that field so they are showing the right things
- It potentially has much higher storage costs (because we’ll store a *lot* more versions)

On the other hand, it’s also a lot more flexible (and safe). ¯\\\_(ツ)\_/¯

/cc @danielballan @weatherpattern @lightandluck 

*File under: this project is slowly becoming a miniaturized archive.org*

### PR DESCRIPTION
**Pull request by Mr0grog**
I guess it’s not like they really conflict.

Fixes #266 ALL THE WAYS

### CODE REVIEW

**Comment by lightandluck** [comment]
Looks good as far as I can tell. Read through code and nothing stuck out. Looked more closely at tests and they make sense and passed. 👍

One minor typo, in the commit message for [355b4212](https://github.com/edgi-govdata-archiving/web-monitoring-db/pull/288/commits/355b421231ba5f18fb01e75e7ce09dd44c7b3ad4), we say the new method is `version.update_changed_attribute`, but it ended up being called `version.update_different_attribute`. Not sure if that's actually worth fixing. ¯\_(ツ)_/¯

File: app/jobs/import_versions_job.rb [hunk 1]
```diff
@@ -174,4 +180,16 @@ def validate_kind!(kinds, record, field_name)
...
     end
   end
+
+  def version_changed?(version)
+    return true if version.version_hash.nil?
+
+    previous = version.page.versions
+      .where(source_type: version.source_type)
+      .where('capture_time < ?', version.capture_time)
+      .order(capture_time: :desc)
```
Comment (danielballan) [inline app/jobs/import_versions_job.rb]: I noticed that, in `update_different_attribute`, you use `limit(1).first` which I assume is a performance win. Does that not buy you anything in this case?

**Comment by lightandluck** [comment]
I went ahead and removed `.limit(1)` and updated the docs, if someone wants to approve and merge =D

**Comment by danielballan** [comment]
Thanks!

**Pull request merged**